master/docs/developer/master-worker-msgpack.rst

```diff
@@ -0,0 +1,1290 @@
+    If ``keepstamp`` was ``True``, workers sends ``update_upload_file_utime`` message.
+    If command succeeded, worker sends ``rc`` value 0 as an ``update`` message ``args`` key-value pair.
+    It can also send ``update`` messages with key ``header`` or ``stderr`` to inform about command execution.
+
+    If command failed, worker sends ``update_upload_file_close`` message and the ``update`` message with dictionary ``args`` key ``rc`` with the error number.
+
+    The basic structure of worker ``update`` message is explained in section :ref:`MsgPack_Keys_And_Values_Message`.
+
+
+Command_name: ``upload_directory``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to ``upload_file``.
+This command will upload an entire directory to the master, in the form of a tarball.
+
+``workdir``
+    Value is a string.
+    It represents a base directory for the filename, relative to the builder's basedir.
+
+``workersource``
+    Value is a string.
+    It represents a path to the worker-side directory to read from, relative to the workdir.
+
+``maxsize``
+    Value is an integer.
+    Maximum number of bytes to transfer from the worker.
+    The operation will fail if the tarball file exceeds this size.
+    Worker will send messages with data to master until it notices it exceeded ``maxsize``.
+
+``blocksize``
+    Value is an integer.
+    Maximum size for each data block to be sent to master.
+
+``compress``
+    Compression algorithm to use – one of ``None``, 'bz2', or 'gz'.
+
+    Worker sends data to the master with one or more ``update_upload_directory_write`` messages.
+    After reading the directory, worker sends ``update_upload_directory_unpack`` with no arguments to extract the tarball and ``rc`` value 0 as an ``update`` message ``args`` key-value pair if the command suceeded.
+
+    Otherwise, worker sends ``update`` message with dictionary ``args`` key ``header`` with information about the error that occured and another ``update`` message with dictionary ``args`` key ``rc`` with the error number.
+
+The basic structure of worker ``update`` message is explained in section :ref:`MsgPack_Keys_And_Values_Message`.
+
+
+Command_name: ``download_file``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Downloads a file from master to worker.
+
+``workdir``
+    Value is a string.
+    It represents a base directory for the filename, relative to the builder's basedir.
+
+``workerdest``
+    Value is a string.
+    It represents a path to the worker-side file to write to, relative to the workdir.
+
+``maxsize``
+    Value is an integer.
+    Maximum number of bytes to transfer from the master.
+    The operation will fail if the file exceeds this size.
+    Worker will request data from master until it notices it exceeded ``maxsize``.
+
+``blocksize``
+    Value is an integer.
+    It represents maximum size for each data block to be sent from master to worker.
+
+``mode``
+    Value is ``None`` or an integer which represents an access mode for the new file.
```

Comment:
I guess we should just refer to unix documentation

Reply:
Buildbot only supports a subset of Unix modes and only those are explained in this documentation.
Referring to full Unix documentation would cause confusion for the user because it would be unclear whan exactly Buildbot supports.